ark/jsonschema/object.ts

```diff
@@ -0,0 +1,336 @@
+		const minProperties = jsonSchema.minProperties
+		predicates.push((data: object, ctx) => {
+			const keys = Object.keys(data)
+			return keys.length >= minProperties ?
+					true
+				:	ctx.reject({
+						expected: `at least ${minProperties} propert${minProperties === 1 ? "y" : "ies"}`,
+						actual: keys.length.toString()
+					})
+		})
+	}
+	return predicates
+}
+
+const parsePatternProperties = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	if (!("patternProperties" in jsonSchema)) return
+
+	const patternProperties = Object.entries(jsonSchema.patternProperties).map(
+		([key, value]) =>
+			[new RegExp(key), innerParseJsonSchema.assert(value)] as const
+	)
+
+	// Ensure that the schema for any property is compatible with any corresponding patternProperties
+	patternProperties.forEach(([pattern, parsedPatternPropertySchema]) => {
+		Object.entries(jsonSchema.properties ?? {}).forEach(
+			([property, schemaForProperty]) => {
+				if (!pattern.test(property)) return
+
+				const parsedPropertySchema =
+					innerParseJsonSchema.assert(schemaForProperty)
+
+				if (!parsedPropertySchema.overlaps(parsedPatternPropertySchema)) {
+					ctx.reject({
+						message: `property ${property} must have a schema that overlaps with the patternProperty ${pattern}`
+					})
+				}
+			}
+		)
+	})
+
+	// NB: We don't validate compatability of schemas for overlapping patternProperties
+	// since getting the intersection of regexes is inherenetly difficult.
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.entries(data).forEach(([dataKey, dataValue]) => {
+			patternProperties.forEach(([pattern, parsedJsonSchema]) => {
+				if (pattern.test(dataKey) && !parsedJsonSchema.allows(dataValue)) {
+					errors.push(
+						ctx.reject({
+							actual: dataValue,
+							expected: `${parsedJsonSchema.description} as property ${dataKey} matches patternProperty ${pattern}`
+						})
+					)
+				}
+			})
+		})
+		return errors.length === 0
+	}
+}
+
+const parsePropertyNames = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	if (!("propertyNames" in jsonSchema)) return
+
+	const propertyNamesValidator = innerParseJsonSchema.assert(
+		jsonSchema.propertyNames
+	)
+
+	if (
+		"domain" in propertyNamesValidator.json &&
+		propertyNamesValidator.json.domain !== "string"
+	) {
+		ctx.reject({
+			path: ["propertyNames"],
+			actual: `a schema for validating a ${propertyNamesValidator.json.domain as string}`,
+			expected: "a schema for validating a string"
+		})
+	}
+
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.keys(data).forEach(key => {
+			if (!propertyNamesValidator.allows(key)) {
+				errors.push(
+					ctx.reject({
+						message: `property ${key} doesn't adhere to the propertyNames schema of ${propertyNamesValidator.description}`
+					})
+				)
+			}
+		})
+		return errors.length === 0
```

Comment:
> In each of these cases, instead of keeping track of the errors with a `false[]`, you can just check `ctx.hasError()` when needed.

I think this is what you meant? de4c7ccf

> Also, by default it's important to try and use `expected` + `actual` for error messages so that they can compose in unions. message is only for the full error with contextualized path which generally should just be handled by the default.
> 
> It can be a bit tricky to word certain errors so they fit in `Must be {expected} (was {actual})`, but there's usually a reasonable way. Try to restructure as many of these errors as possible to use those.

Done in aa817827